[SPARK-18091] [SQL] [BACKPORT-1.6] Deep if expressions cause Generated SpecificUnsafeProjection code to exceed JVM code size limit

kapilsingh5050 opened the pull request:
## What changes were proposed in this pull request?

Fix for SPARK-18091 which is a bug related to large if expressions causing generated SpecificUnsafeProjection code to exceed JVM code size limit.

This PR changes if expression's code generation to place its predicate, true value and false value expressions' generated code in separate methods in context so as to never generate too long combined code.

## How was this patch tested?

Added a unit test and also tested manually with the application (having transformations similar to the unit test) which caused the issue to be identified in the first place.

Commit:
[SPARK-14578] [SQL] [BACKPORT-1.6] Fix codegen for CreateExternalRow with nested wide schema

The wide schema, the expression of fields will be splitted into multiple functions, but the variable for loopVar can't be accessed in splitted functions, this PR change them as class member.

Added regression test.

Author: Davies Liu <davies@databricks.com>

Closes #12338 from davies/nested_row.

Conflicts:
	sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala

Commit:
Fixed If expression code generation to never generate code large enough to exceed JVM method code size limit

AmplabJenkins commented:
Can one of the admins verify this patch?

cloud-fan commented:
ok to test

SparkQA commented:
**[Test build #69707 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69707/consoleFull)** for PR 16146 at commit [`8672343`](https://github.com/apache/spark/commit/86723436ba2b711d0eb6f2de92f3651006e3bff4).

SparkQA commented:
**[Test build #69707 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69707/consoleFull)** for PR 16146 at commit [`8672343`](https://github.com/apache/spark/commit/86723436ba2b711d0eb6f2de92f3651006e3bff4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69707/
Test FAILed.

cloud-fan committed:
[SQL][MINOR] simplify a test to fix the maven tests

## What changes were proposed in this pull request?

After https://github.com/apache/spark/pull/15620 , all of the Maven-based 2.0 Jenkins jobs time out consistently. As I pointed out in https://github.com/apache/spark/pull/15620#discussion_r91829129 , it seems that the regression test is an overkill and may hit constants pool size limitation, which is a known issue and hasn't been fixed yet.

Since #15620 only fix the code size limitation problem, we can simplify the test to avoid hitting constants pool size limitation.

## How was this patch tested?

test only change

Author: Wenchen Fan <wenchen@databricks.com>

Closes #16244 from cloud-fan/minor.

cloud-fan commented:
hi @kapilsingh5050  can you also include https://github.com/apache/spark/pull/16244? It does fix the maven tests.

kapilsingh5050 commented:
Yes, I'll do that but the test failures here are different. I'm still to figure out the root cause.

Commit:
[minor] Fixed unit test as per 1.6

kapilsingh5050 commented:
ExpressionEncoderSuite and RowEncoderSuite tests were failing because of following fix missing in branch-1.6:
https://github.com/apache/spark/commit/372baf0479840695388515170e6eae0b3fc4125e

SparkQA commented:
**[Test build #70015 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70015/consoleFull)** for PR 16146 at commit [`73f1231`](https://github.com/apache/spark/commit/73f1231a37ccf9ec9be338072803a71bfff41726).

SparkQA commented:
**[Test build #70015 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70015/consoleFull)** for PR 16146 at commit [`73f1231`](https://github.com/apache/spark/commit/73f1231a37ccf9ec9be338072803a71bfff41726).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70015/
Test FAILed.

SparkQA commented:
**[Test build #70185 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70185/consoleFull)** for PR 16146 at commit [`708c847`](https://github.com/apache/spark/commit/708c847bd86ded2f58ec0ff05424cc437aac99b5).

SparkQA commented:
**[Test build #70185 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70185/consoleFull)** for PR 16146 at commit [`708c847`](https://github.com/apache/spark/commit/708c847bd86ded2f58ec0ff05424cc437aac99b5).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70185/
Test FAILed.

kapilsingh5050 commented:
@cloud-fan The last test failure is because of following error:
"impossible to get artifacts when data has not been loaded. IvyNode = org.scala-lang#scala-library;2.10.3"
for which [this](https://github.com/sbt/sbt/issues/2015) post suggests changing sbt version to 0.13.9 and I also found an open SPARK bug corresponding to this error. 
Are you aware of any recent changes in the build configuration for PR builder?
Can we retry the build?

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #70189 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70189/consoleFull)** for PR 16146 at commit [`708c847`](https://github.com/apache/spark/commit/708c847bd86ded2f58ec0ff05424cc437aac99b5).

SparkQA commented:
**[Test build #70189 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70189/consoleFull)** for PR 16146 at commit [`708c847`](https://github.com/apache/spark/commit/708c847bd86ded2f58ec0ff05424cc437aac99b5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70189/
Test PASSed.

kapilsingh5050 commented:
@cloud-fan please review and merge

rezasafi commented:
Hi, will this pull request merge to branch-1.6?

zzcclp commented:
ping @cloud-fan , will this pr be merged into branch-1.6?
